File: 0041-first-missing-positive/0041-first-missing-positive.go
func firstMissingPositive(nums []int) int {
    n := len(nums)
    m := 0
    for i:=0; i < n; i++ {
        if nums[i] < 0 {
            nums[i] = 0
        } else if nums[i] > m {
            m = nums[i]
        }
    }

    m++;
    for _, num := range nums {
        orig := num
        if orig >= m {
            orig -= m
        }

        if orig > 0 && orig <= n && nums[orig - 1] < m {
            nums[orig - 1] += m
        }

    }

    for i:=0; i < n; i++ {
        if nums[i] < m {
            return i + 1
        }
    }

    return n + 1
}

File: 0041-first-missing-positive/0041-first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            if nums[i] < 0:
                nums[i] = 0

        m = max(nums)
        m += 1
        for i in range(len(nums)):
            orig = nums[i]
            if orig >= m:
                orig -= m

            if 0 < orig <= len(nums) and nums[orig - 1] < m:
                nums[orig - 1] += m

        for i in range(len(nums)):
            if nums[i] < m:
                return i + 1

        return len(nums) + 1
